File: LeetCode/cpp/1190_Reverse_Substrings_Between_Each_Pair_of_Parentheses.cpp
class Solution {
public:
	string reverseParentheses(string s) {
		string answer;
		stack<char> st;

		string temp;
		for (int i = 0; i < s.length(); i++) {
			if (s[i] == ')') {
				while (!st.empty() && st.top() != '(') {
					temp += st.top();
					st.pop();
				}
				st.pop();

				for (int j = 0; j < temp.length(); j++) {
					st.push(temp[j]);
				}
				temp.clear();
			}
			else {
				st.push(s[i]);
			}
		}

		while (!st.empty()) {
			answer += st.top();
			st.pop();
		}

		reverse(answer.begin(), answer.end());

		return answer;
	}
};

File: LeetCode/java/1190_Reverse_Substrings_Between_Each_Pair_of_Parentheses.java
import java.util.Stack;

class Solution {
    public String reverseParentheses(String s) {
        StringBuilder answer = new StringBuilder();;
        Stack<Character> stack = new Stack<>();

        StringBuilder temp = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == ')') {
                while (!stack.isEmpty() && stack.peek() != '(') {
                    temp.append(stack.pop());
                }
                stack.pop();

                for (int j = 0; j < temp.length(); j++) {
                    stack.push(temp.charAt(j));
                }
                temp.setLength(0);
            } else {
                stack.push(s.charAt(i));
            }
        }

        while (!stack.isEmpty()) {
            answer.append(stack.pop());
        }

        answer.reverse();

        return answer.toString();
    }
}